common/src/main/java/io/netty/util/collection/PrimitiveCollections.java

```diff
@@ -171,42 +189,80 @@ public boolean containsKey(int key) {
         }
 
         @Override
-        public boolean containsValue(V value) {
+        public boolean containsValue(Object value) {
             return map.containsValue(value);
         }
 
         @Override
-        public Iterable<Entry<V>> entries() {
-            return this;
+        public boolean containsKey(Object key) {
+            return map.containsKey(key);
+        }
+
+        @Override
+        public V get(Object key) {
+            return map.get(key);
+        }
+
+        @Override
+        public V put(Integer key, V value) {
+            throw new UnsupportedOperationException("put");
+        }
+
+        @Override
+        public V remove(Object key) {
+            throw new UnsupportedOperationException("remove");
         }
 
         @Override
-        public Iterator<Entry<V>> iterator() {
-            return new IteratorImpl(map.entries().iterator());
+        public void putAll(Map<? extends Integer, ? extends V> m) {
+            throw new UnsupportedOperationException("putAll");
         }
 
         @Override
-        public int[] keys() {
-            return map.keys();
+        public Iterable<PEntry<V>> entries() {
+            if (entries == null) {
+                entries = new Iterable<PEntry<V>>() {
+                    @Override
+                    public Iterator<PEntry<V>> iterator() {
+                        return new IteratorImpl(map.entries().iterator());
+                    }
+                };
+            }
+
+            return entries;
         }
 
         @Override
-        public V[] values(Class<V> clazz) {
-            return map.values(clazz);
+        public Set<Integer> keySet() {
+            if (keySet == null) {
+                keySet = Collections.unmodifiableSet(map.keySet());
+            }
+            return keySet;
+        }
+
+        @Override
+        public Set<Entry<Integer, V>> entrySet() {
+            if (entrySet == null) {
+                entrySet = Collections.unmodifiableSet(map.entrySet());
+            }
+            return entrySet;
         }
 
         @Override
         public Collection<V> values() {
-            return map.values();
+            if (values == null) {
+                values = Collections.unmodifiableCollection(map.values());
+            }
+            return values;
         }
 
         /**
          * Unmodifiable wrapper for an iterator.
          */
-        private class IteratorImpl implements Iterator<Entry<V>> {
-            final Iterator<Entry<V>> iter;
+        private class IteratorImpl implements Iterator<PEntry<V>> {
+            final Iterator<PEntry<V>> iter;
```

Comment:
@nmittler - I'm not sure what we are gaining by abbreviating here. Is `Primitive` overly verbose?